opm/simulators/flow/FlowProblemComp.hpp

```diff
@@ -438,13 +449,43 @@ class FlowProblemComp : public FlowProblem<TypeTag>
                     dofFluidState.setPressure(phaseIdx, pressure);
             }
 
-            for (unsigned compIdx = 0; compIdx < numComponents; ++compIdx) {
-                const std::size_t data_idx = compIdx * numDof + dofIdx;
-                const Scalar xmf = xmfData[data_idx];
-                const Scalar ymf = ymfData[data_idx];
+            if (has_xmf && has_ymf) {
+                const auto& xmfData = fp.get_double("XMF");
+                const auto& ymfData = fp.get_double("YMF");
+                for (unsigned compIdx = 0; compIdx < numComponents; ++compIdx) {
+                    const std::size_t data_idx = compIdx * numDof + dofIdx;
+                    const Scalar xmf = xmfData[data_idx];
+                    const Scalar ymf = ymfData[data_idx];
+
+                    dofFluidState.setMoleFraction(FluidSystem::oilPhaseIdx, compIdx, xmf);
+                    dofFluidState.setMoleFraction(FluidSystem::gasPhaseIdx, compIdx, ymf);
+                }
+            }
 
-                dofFluidState.setMoleFraction(FluidSystem::oilPhaseIdx, compIdx, xmf);
-                dofFluidState.setMoleFraction(FluidSystem::gasPhaseIdx, compIdx, ymf);
+            if (has_zmf) {
+                const auto& zmfData = fp.get_double("ZMF");
+                for (unsigned compIdx = 0; compIdx < numComponents; ++compIdx) {
+                    const std::size_t data_idx = compIdx * numDof + dofIdx;
+                    const Scalar zmf = zmfData[data_idx];
+                    // TODO: I know we do this for co2_ptflash example, but maybe we should not do it this way?
+                    if (gas_active) {
+                        const auto& sat_gas = dofFluidState.saturation(gasPhaseIdx);
+                        if (sat_gas > 0.) {
+                            dofFluidState.setMoleFraction(FluidSystem::gasPhaseIdx, compIdx, zmf);
+                        } else {
+                            dofFluidState.setMoleFraction(FluidSystem::gasPhaseIdx, compIdx, 0.);
+                        }
+                    }
+                    if (oil_active) {
+                        dofFluidState.setMoleFraction(FluidSystem::oilPhaseIdx, compIdx, zmf);
+                        const auto& sat_oil = dofFluidState.saturation(oilPhaseIdx);
+                        if (sat_oil > 0.) {
+                            dofFluidState.setMoleFraction(FluidSystem::oilPhaseIdx, compIdx, zmf);
+                        } else {
+                            dofFluidState.setMoleFraction(FluidSystem::oilPhaseIdx, compIdx, 0.);
+                        }
```

Comment:
Does `setMoleFraction()` affect the return value of `saturation()`?  If not, the `sat_oil > 0` branch is redundant.  Maybe this could be rewritten as
```C++
dofFluidState.setMoleFraction(FluidSystem::oilPhaseIdx, compIdx, zmf);
if (! (dofFluidState.saturation(oilPhaseIdx) > 0)) {
    dofFluidState.setMoleFraction(FluidSystem::oilPhaseIdx, compIdx, Scalar{0});
}
```
or maybe even
```C++
const auto xmf = (dofFluidState.saturation(oilPhaseIdx) > 0) ? zmf : Scalar{0};
dofFluidState.setMoleFraction(FluidSystem::oilPhaseIdx, compIdx, xmf);
```
By the way, this block uses both the name `FluidSystem::oilPhaseIdx` and the unadorned name `oilPhaseIdx`.  Do those two the represent the same index value?

Reply:
> Does setMoleFraction() affect the return value of saturation()? If not, the sat_oil > 0 branch is redundant. Maybe this could be rewritten as

No, they are different things. Here, I want to set mole fraction is 0 for the phase does not exist, for example, the cell does not have oil, I will set the mole fraction in oil 0.  So I should have removed the following line, 
```c++
                        dofFluidState.setMoleFraction(FluidSystem::oilPhaseIdx, compIdx, zmf);
```

Reply:
> 
```
const auto xmf = (dofFluidState.saturation(oilPhaseIdx) > 0) ? zmf : Scalar{0};
dofFluidState.setMoleFraction(FluidSystem::oilPhaseIdx, compIdx, xmf);
```

I will use this one. 

`oilPhaseIdx` is the same with `FluidSystem::oilPhaseIdx`,  defined in FlowProblem.hpp with ` enum { oilPhaseIdx = FluidSystem::oilPhaseIdx };`